minimal scala native support - resolves ##639

rwhaling opened the pull request:
This provides support for Scala Native in zio core, utilizing our libuv executionContext scala-native-loop, and a compatability shim for the scalajs timer API.  Since we are line-for-line compatible with Scala.js for the purpose of Zio, I've imported the scalajs sources into the native subproject - no new code.

Open questions are -
1. Is it wise to implement just core and stacktracer to start, or should we try to cover more of the subprojects before merging?
2. Would it be better to copy/paste the scala.js platform code?
3. This published local artifacts fine with `coreNative/publishLocal` and `stacktracerNative/publishLocal`, but thats probably not enough for CI, right?

jdegoes commented:
@rwhaling Excited to get this merged! Can you run `fmt` to get pass the linter?

Meanwhile, any ZIO contributors / users familiar with Native and willing to give this a review / spin?

neko-kai commented:
briefly hijacked for fmt

neko-kai commented:
I guess we should wait until all tests migrate to ZIO Test since we can't (seemingly ?) run current tests

rwhaling commented:
thanks @neko-kai! Let me see what I can do about the tests.

adamgfraser commented:
Is there anything we need to do as we build out ZIO Test to make sure we can support Scala Native as well? Or since you are compatible with Scala.js are we good as long as we can support that?

rwhaling commented:
@adamgfraser - that's a really interesting question.  Testing on Native is kind of fiddly and a pain, since it is so far from what sbt expects - let me consult with a few other SN contributors and get back to you?

adamgfraser commented:
@rwhaling That would be great! We currently have an SBT runner but specifications are also their own executable programs. So if SBT was the issue you wouldn't necessarily need to use SBT to use ZIO Test. You can try doing `test:run` in the `coreTestsJVM` project to see what that looks like.

sideeffffect commented:
this is no longer blocked by https://github.com/zio/zio/issues/1208

also, this issue deserves the [`scala-native` badge](https://github.com/zio/zio/issues?q=label%3Ascala-native)

sideeffffect commented:
@rwhaling the PR is green and approved. Only some conflicts with sbt remain. Can you resolve them? Or would you like me to do that?

rwhaling commented:
@sideeffffect thanks for nudging this along, this PR fell off my radar a bit; have also been triaging some issues that affect this implementation over in SN.  

I've also seen a few quirks from the way I am re-using the JS tree verbatim here:
https://github.com/zio/zio/pull/1428/files#diff-fdc3abdfd754eeb24090dbd90aeec2ceR92

It looks like it is pulling in the wrong executionContext because of this, which would break a lot of things in a more complex program.

What I propose to do is fix the SBT conflicts, but also add a proper implementation of coreNative - it would be 90% identical to JS, but I think it'll be better in the long-run. Does that sound like a good plan to you?

rwhaling commented:
@sideeffffect I can take a first pass over the weekend - if you can review and help get it over the finish line, it shouldn't be a ton of work for either of us.

rwhaling commented:
@sideeffffect - I updated my PR, fixed the merge conflicts, but now it won't compile - specs2 4.8.0 isn't published for native yet: https://mvnrepository.com/artifact/org.specs2/specs2-core

Writing up an issue now so we can get this merged.

adamgfraser commented:
@rwhaling We're in the process of migrating away from Spec2 to ZIO Test and there are currently only two files that use Specs 2. We have an open ticket for that in #1647 so one solution is to just complete that migration.

But in advance of that would it be possible to just move those two files to a separate directory and include settings in `BuildHelper.scala` to only compile those two test files on JVM / ScalaJS? Then we could alter the `build.sbt` settings so that Specs2 was only included as a dependency for JVM and ScalaJS. Since it is only two files this would not result in much loss in test coverage and would let us push forward with this initiative immediately.

What do you think?

rwhaling commented:
@adamgfraser I'm definitely in favor of moving away from fragile dependencies.  

I like your idea a lot, but I'm probably not confident enough in my understanding of specs2 or the Zio build to execute it.  What do you think @sideeffffect?

We did get unblocked by specs2, that artifact should be up on Maven Central in an hour or two, but tbh it is likely to fail again.  

Depending on how long you think it will take to migrate those last two files, we can either merge as-is once we can get CI to pass (hopefully later today), or do the refactor if y'all think it is a quick win - but I've had this PR out for soooo long I probably want to merge as soon as is feasible.

adamgfraser commented:
I can try to help with the build tonight. Not sure what time zone you’re in. We have a ticket for those two tests but I don’t think anyone is working on them right now, so we shouldn’t wait for that. I’m coming into this a little from the outside so I am fine with whatever you decide to do but my experience from Dotty is there can be some weird corner cases so it would be nice if we could get most of the tests running so we could have some confidence that things were working properly.

adamgfraser commented:
@rwhaling We migrated the last tests to ZIO Test and just removed the Specs2 dependency so the way should be clear to merge this. Can you merge upstream changes and resolve any conflicts? Let me know if you need any help. Will be great to have this in for the 1.0.0 release!

sideeffffect commented on build.sbt:
```diff
@@ -199,7 +210,11 @@ lazy val stacktracerJVM = stacktracer.jvm
+  .settings(scalaVersion := "2.11.12")
+  .settings(skip in Test := true)
+  .settings(skip in doc := true)
+  lazy val testRunner = crossProject(JVMPlatform, JSPlatform)
```
unnecessarily changed indentation

sideeffffect commented on core/native/src/main/scala/zio/internal/OneShot.scala:
```diff
@@ -0,0 +1,55 @@
+   * Retrieves the value of the variable, blocking if necessary.
+   */
+  final def get(): A = {
+    if (value == null) throw new Error("Cannot block for result to be set in JavaScript")
```
this references "JavaScript"

rwhaling replied:
This might still be worth changing to Native

sideeffffect replied:
or `native-loop`?

sideeffffect commented on core/native/src/main/scala/zio/scheduler/SchedulerLive.scala:
```diff
@@ -0,0 +1,75 @@
+import zio.duration.Duration
+import zio.internal.{ Scheduler => IScheduler }
+
+import scala.scalajs.js
```
is it correct to import `scalajs` stuff here? (or is that what `native-loop-js-compat` is for?)

rwhaling replied:
@sideeffffect - yep, exactly.  We have a compatability shim that presents Scala Native's async capabilities with a scalajs api.

rwhaling commented:
@adamgfraser @sideeffffect - I should be able to do another pass Wednesday or Thursday, thanks for your patience and persistence 😄

adamgfraser commented:
@rwhaling Thank you!

rwhaling commented:
@adamgfraser @sideeffffect - OK, up-to-date now and all tests passing.  Ready to merge?

sideeffffect commented on build.sbt:
```diff
@@ -95,6 +95,18 @@ lazy val coreJVM = core.jvm
+  .settings(scalaVersion := "2.11.12")
+  .settings(skip in Test := true)
+  .settings(skip in doc := true)
+  .settings(sources in (Compile, doc) := Seq.empty)
```
maybe
```scala
Compile / doc / sources
```
(and similarly in other places)
but this is a trivial detail...

contrun commented on core/native/src/main/scala/zio/internal/Sync.scala:
```diff
@@ -0,0 +1,8 @@
+package zio.internal
+
+private[zio] object Sync {
```
Can someone please explain the use of this object? I am really baffled.
